Adds "How to verify contracts" to the How To section with examples (#578)
matter-labs/zksync-web-era-docs: docs/dev/how-to/verify-contracts.md

Contract source-code verification ensures that the code running on-chain matches your published code. 

The verification process validates and authenticates contracts running on a blockchain network, and enhances transparency, security, and trust in your smart contracts.

This document shows you how to verify your contracts with the [`hardhat-zksync-verify`](https://www.npmjs.com/package/@matterlabs/hardhat-zksync-verify) plugin.

## Common use cases

### Transparent contract deployment

Using the plugin, developers can deploy their contracts on the zkSync Era network with transparency so that users and other developers can independently verify the contract's source code, and ensure it behaves as expected.

### Cross-chain interoperability

For contracts operating across multiple chains, verifying contracts on each network assures users of the contract's consistency across different networks.

### Open source projects

For open-source projects, verifying contracts enhances trust and encourages more developers to contribute. It assures contributors that the deployed contracts match the source code.
  
## Verifying contracts using `hardhat-zksync-verify`

> ### 1. Project Setup

Reviewer: ```suggestion
### 1. Project setup
```

> 1. Begin by installing [zkSync CLI](/docs/tools/zksync-cli/) to establish a new project:

Reviewer: Better to use bullets as subdivisions of numbers.

```sh
yarn add global zksync-cli@latest
```

> 2. Once you complete the installation, execute the command below to create a fresh project:

Reviewer: ```suggestion
- Once you complete the installation, execute the command below to create a fresh project:
```

```sh
yarn zksync-cli create verify-greeter-contract
```

3. This function creates a fresh zkSync project titled `verify-greeter-contract` containing a `Greeter` contract. Proceed by moving into the project directory:

```sh
cd verify-greeter-contract
```

> ### 2. Package Installation for Verification

Reviewer: ```suggestion
### 2. Package installation for verification
```

        ethNetwork: "http://localhost:8545",
        zksync: true,
      }
    : {
        url: "https://zksync2-testnet.zksync.dev",
        ethNetwork: "goerli",
        zksync: true,
        verifyURL: 'https://zksync2-testnet-explorer.zksync.dev/contract_verification'  // Verification endpoint
      };

const config: HardhatUserConfig = {
  zksolc: {
    version: "1.3.10",
    compilerSource: "binary",
    settings: {},
  },
  defaultNetwork: "zkSyncTestnet",
  networks: {
    hardhat: {
      zksync: false,
    },
    zkSyncTestnet,
  },
  // etherscan: { // Optional - If you plan on verifying a smart contract on Ethereum within the same project
  //   apiKey: //<Your API key for Etherscan>,
  // },
  solidity: {
    version: "0.8.17",
  },
};

export default config;
```

- Feel free to assign an arbitrary zkSync Era network name utilizing the `defaultNetwork` property. 
- The `verifyURL` attribute directs to the verification endpoint specific to the zkSync network.
- If you intend to validate a smart contract on Ethereum within the same project, don't forget to include your [Etherscan API key](https://docs.etherscan.io/getting-started/viewing-api-usage-statistics).

> ### 4. Greeter Contract Compilation

Reviewer: ```suggestion
### 4. Greeter contract compilation
```

The [zkSync CLI](/docs/tools/zksync-cli/) provides a `Greeter.sol` contract we will verify on zkSync Era. 

Compile the contract using this command:

```sh
yarn hardhat compile
```

> ### 5. Deployment of the Greeter Contract 

Reviewer: ```suggestion
### 5. Deploy the Greeter contract 
```

> The [zkSync CLI](/docs/tools/zksync-cli/) provides a `deploy/deploy-greeter.ts` script that we will use to deploy the Greeter contract. Replace `<WALLET-PRIVATE-KEY>` with your key.

Reviewer: ```suggestion
The [zkSync CLI](/docs/tools/zksync-cli/) provides a `deploy/deploy-greeter.ts` script that we will use to deploy the Greeter contract. 

Add your private key to `<WALLET-PRIVATE-KEY>` in the `.env.example` file and remove `.example`.

```

Initiate contract deployment using this command:

```sh
yarn hardhat deploy-zksync --script deploy-greeter.ts
```

Expect an output similar to this:

```txt
Running the deployment function for the Greeter contract
The deployment is estimated to cost 0.0265726735 ETH
constructor args:0x000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000000094869207468657265210000000000000000000000000000000000000000000000
The Greeter contract got deployed at 0xE84774C41F096Ba5BafA1439cEE787D9dD1A6b72
```

> Remember, you will need the contract address to verify the contract on zkSync Era. 

Reviewer: ```suggestion
Remember, you need the contract address to verify the contract on zkSync Era. 
```

### 6. Verify the contract

Run the following command to verify your contract on the specified network, replacing `<contract address>` with your contract's actual address.

```sh
yarn hardhat verify --network <network> <contract address>
```

For example, to verify our Greeter contract on the `zkSyncTestnet` network and our contract address is `0xE84774C41F096Ba5BafA1439cEE787D9dD1A6b72`, use:

```sh
yarn hardhat verify --network zkSyncTestnet 0xE84774C41F096Ba5BafA1439cEE787D9dD1A6b72 'Hi There!'

> ```

Reviewer: I get errors here because there is no "Hi There!" in the deploy, I'm guessing. 

With Hi there, copy/pasted, I get:

```
Error in plugin @matterlabs/hardhat-zksync-verify: Constructor arguments are not correct
```

With no constructor I get:

```
Error in plugin @matterlabs/hardhat-zksync-verify: The number of constructor arguments you provided (0) does not match the number of constructor arguments the contract has been deployed with (1).
```

Author: My mistake, the args in the docs should be 'Hi there!' instead of 'Hi There!'. Updated in latest commit.

```typescript
const contractAddress = "<your contract address>";
const contractFullyQualifiedName = "<your contract fully qualified name>";
const constructorArguments = [/* your decoded constructor arguments */];

const verificationId = await hre.run("verify:verify", {
  address: contractAddress,
  contract: contractFullyQualifiedName,
  constructorArguments: constructorArguments
});

console.log(`Verification ID: ${verificationId}`);
```

In this script:

- `contractAddress` is the address of the deployed contract you wish to verify. 
- `contractFullyQualifiedName` is the fully qualified name of your contract (including the path to your Solidity file and contract name, separated by a colon e.g., `"contracts/Greeter.sol:Greeter"`).
- `constructorArguments` is an array containing the arguments used to deploy your contract (e.g. "Hi There").

Once this script runs, it prints the verification ID. If the verification request is successful, you can use this ID to check the status of your verification request. If the request was not successful, the return value and printed ID is `-1`.

> Contract verification in zkSync Era ensures the integrity and trustworthiness of your contracts. The [zkSync Era Block Explorer](https://explorer.zksync.io/blocks/) is a manual UI process for doing the same, ideal for occasional use, while the `hardhat-zksync-verify` plugin facilitates an automated, flexible approach for developers.

Reviewer: ```suggestion
9. UI block explorer alternative

Contract verification in zkSync Era ensures the integrity and trustworthiness of your contracts. The [zkSync Era Block Explorer](https://explorer.zksync.io/blocks/) is a manual UI process for doing the same, ideal for occasional use, while the `hardhat-zksync-verify` plugin facilitates an automated, flexible approach for developers.
```
